### PR TITLE
Timing model composition

### PR DESCRIPTION
**Pull request by luojing1211**
@ALL, 

This is PR changes the current timingmodel style to composition style with the ability to handle the component order from the model builder. It is fully tested with python 2.7 and python 3. 

Adding removing components are simplified a lot. But not in the way of making a new TimingModel class.

### CODE REVIEW

**Comment by luojing1211** [comment]
@jellis18 could you check what is wrong with Python 3? Why it is failing

**Comment by jellis18** [comment]
@luojing1211, I'm not 100% sure why it is failing but I think that you may need to change the `getattr` method in `TimingModel` to do a  check like that used in the `__new__`  method of `observatory.py`.

File: pint/models/model_builder.py [hunk 1]
```diff
@@ -129,54 +90,63 @@ def preprocess_parfile(self,parfile):
         pfile.close()
         return self.param_inparF
 
+    def get_all_categroies(self,):
```
Comment (paulray) [inline pint/models/model_builder.py]: categories is misspelled here

**Comment by luojing1211** [comment]
I have updated the adding and removing a component part.  It is not in the way of making new TimingModel class. But replace the orderedDict with a new ordered list of components. This way we don't have to worry about if the information has been copied to the new TimingModel class.

Please let me know is this a better way to do it. 
If we agree this is the way. We can merge it.

**Comment by paulray** [comment]
What is preventing this from being merged now? I'd like it merged so I could add some changes on top of it without making conflicts. @luojing1211 has requested reviews from @scottransom @demorest and @jellis18. Are you OK with merging?

**Comment by luojing1211** [comment]
The parts need to be reviewed is add_components, change_component_order section. If it is ok for you, the reviewers, we can start merge it. Future improvement can be done by smaller PR.

**Comment by scottransom** [comment]
Maybe you can walk us through the add_components and change_component_order code tomorrow?
One thing that kind of worries me is that I think we thought that this would simplify the code.  But it appears as if there are a lot more new lines rather than lines removed -- and that is even given that you are deleting a bunch of cache lines!  Any idea why that is?

**Comment by luojing1211** [comment]
It looks ok now.

**Pull request merged**